datafusion/src/arrow_temporal_util.rs

```diff
@@ -0,0 +1,302 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```

Comment:
This file is a copy of `arrow/src/compute/kernels/cast_utils.rs`.

Arrow2 currently only provide `utf8_to_timestamp_ns_scalar` with explicit timezone and fmt string, or `utf8_to_naive_timestamp_ns_scalar` with explicit fmt string. It's not possible to achieve identical semantic with the original arrow-rs `string_to_timestamp_nanos ` in arrow2 since  (as I quote jorge here)

> it mixes timezone-aware, non-timezone aware, and local time in the same conversion. That IMO is not a “cast”: it is a “let me try anything to not crash irrespectively of its correctness”. For people that do rely on timezones, imo that is a foot gun that hides bugs. It also goes against the arrow spec that states that naive timestamps are not utc.

I added this back as a temporary workaround to make the build pass. We need to redesign the time-related API by providing users with two possible flavors: one with explicit timezone setting and the other default to local timezone, but both of the API variants should ask user for datetime format string.  I'll file an issue on this later.

Reply:
i think it's the right thing to move this logic into datafusion.